### DISCUSSION TITLE
Restful  in avalonia browser program

### DISCUSSION THREAD

**Discussion by zy1075984** [AvaloniaUI/Avalonia#11981] [Q&A]
I used restsharp to develop a restful client program, the desktop program is used normally, the browser program to send a request, the server has received, but blocked in the client Get method, what is the solution? The code is as follows:
`        var client = new RestClient("http://localhost:8021/");
        var request = new RestRequest("GetInput/home_timeline");
        var response = client.Get(request);`

I am a novice in web development. Can you provide an example of how to use it?

**Comment by hez2010** [comment]
I see you are doing synchronized requests here. In the browser, keep in mind that you need to do async calls for every I/O request to avoid deadlock, because the browser environment is single threaded so sync over async things simply won't work at all. 

So you need to use `var response = await client.GetAsync(request)` instead of `var response = client.Get(request);`.

**Reply by zy1075984** [reply]
Thank you for your reply, I tried to use await or task, no blocked will occur, but still not get the data

**Reply by hez2010** [reply]
How about using HttpClient directly?

**Reply by hhgm** [reply]
I have an application pre-version 11 that uses RestSharp... and that works fine.. but in version 11, I am not able to use restsharp.. on Android it crashes when I instantiate a client, and in desktop I get all the calls terminated .. I realize Avalonia 8 is based on dotnet8, but is that the answer? RestSharp is not compatible, or are there some other explanation?

**Reply by thevortexcloud** [reply]
What's the error you are getting?

**Comment by hhgm** [comment]
My code:

>             Parallel.ForEach(segment?.Hosts(), new ParallelOptions { MaxDegreeOfParallelism = 4 }, (entry) =>
            //var hosts = segment?.Hosts();
            //foreach (var host in hosts)
            {
                //Task.Factory.StartNew(() =>
                //{
                string ipAddress = entry.ToIpString();
                _logger.Debug($"Discover: {ipAddress}");

                string endpoint = "api/info/details";

                try
                {
                    //var restClient = new RestClient($"https://{ipAddress}:{_port}/");
                    RestClient restClient = new RestClient($"https://{ipAddress}:{_port}/", configureSerialization: s => s.UseNewtonsoftJson(), configureRestClient: options => options.RemoteCertificateValidationCallback = RemoteCertificateValidationCallback);

                    Debug.WriteLine($"Detecting {ipAddress}");
                    RestRequest request = new RestRequest(endpoint);
                    request.Timeout = 3000;
                    request.RequestFormat = DataFormat.Json;
                    RestResponse<DeviceDetails> response =
                        restClient.Execute<DeviceDetails>(request);

                    if (response.IsSuccessful)
                    {
                        _logger.Debug($"Satellite found on: {ipAddress}");
                        //we found a satellite
                        SatelliteDiscovered?.Invoke(this, new SatelliteDiscoveredEventArguments(response.Data));
                    }
                    else
                    {
                        _logger.Info($"IP:{ipAddress}, Message: {response.ErrorMessage}");
                    }

                    Avalonia.Threading.Dispatcher.UIThread.Post(() =>
                    {
                        DiscoverProcessed++;
                    });
                }
                catch (Exception ex)
                {
                    _logger.Error(ex);
                }
            }

So, I actually dug a bit deeper, and it seems that my certification validation callback is never reached.. 
Thus, I get the error: {"No connection could be made because the target machine actively refused it. (192.168.0.128:8080)"} in the RestSharp response.

**Reply by thevortexcloud** [reply]
That sort of error typically means the server is not actually up or there is a firewall stopping it. Have you tried testing if you can connect in a web browser? Have you also tested raw HttpClient? EG 

```csharp
var client  = new HttpClient();
var response = await client.GetStringAsync("https://192.168.0.128:8080/api/info/details");
Console.WriteLine(response);
```

On an unrelated note, your code is coupling a lot of unrelated things together. It's typically considered to be good practice to separate your data access from your presentation.

You are also instantiating a new client per host (at least that's what I assume RestSharp is doing). This is bad. You should make a single client and reuse it. Otherwise you may get socket exhaustion.

https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines

**Reply by maxkatz6** [reply]
Also, remember that CORS exists in Browsers. https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

**Reply by hhgm** [reply]
Thank you...right now just quick and dirty code.. not polished in any way.. I will try out later today to see what the core issue is.

**Reply by thevortexcloud** [reply]
> right now just quick and dirty code.. not polished in any way.. 

All good. I have seen a lot of real apps out in the wild that work like this so I can never be sure if someone is doing a quick test or intends to stick with it.

**Reply by hhgm** [reply]
Right now I am just trying to get the basic functionalities to work... then I will clean it up and do it proper.. so much more to it that this..  :-)
